### ISSUE TITLE
ExpectedExceptionSupport causes exceptions to be silently ignored

### ISSUE THREAD

**Issue by algielen** [junit-team/junit5#1069]
## Overview

When adding the class annotation `@EnableRuleMigrationSupport`, no exceptions thrown will make the tests fail.

Example code :

```java
package be.algielen.soap;

import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.migrationsupport.rules.EnableRuleMigrationSupport;

@EnableRuleMigrationSupport
class HelloServiceTest {

    @Test
    void sayHello() {
        assertTrue(true);
    }

    @Test
    void throwsException() {
        throw new RuntimeException();
    }
}
```

All tests succeed even though `throwsException()` should obviously fail.

Version I tested it on : 

```xml
<plugin>
        <artifactId>maven-surefire-plugin</artifactId>
        <version>2.19</version>

        <configuration>
          <argLine>-Xmx512M</argLine>
        </configuration>
        <dependencies>
          <dependency>
            <groupId>org.junit.platform</groupId>
            <artifactId>junit-platform-surefire-provider</artifactId>
            <version>1.0.0</version>
            <scope>compile</scope>
          </dependency>
          <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-engine</artifactId>
            <version>5.0.0</version>
            <scope>compile</scope>
          </dependency>
        </dependencies>
      </plugin>
...
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-api</artifactId>
      <version>5.0.0</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-migrationsupport</artifactId>
      <version>5.0.0</version>
      <scope>test</scope>
    </dependency>
```

Using Maven surefire 2.19 and with the integrated JUnit 5 Runner of Intellij IDEA 2012.2.4

## Deliverables

- [ ] `ExpectedExceptionSupport` must never swallow unexpected exceptions thrown from test methods.
- [ ] Document fix in the release notes in the User Guide.

**Comment by sbrannen** [comment]
Good catch! 👍 

We aim to fix this bug this in JUnit Jupiter 5.1.

**Comment by sbrannen** [comment]
@junit-team/junit-lambda This is technically the first candidate issue for a 5.0.1 release.

Though perhaps we should wait to see how many other bugs are reported in the coming weeks/months before deciding to put out a point release.

Thoughts?

**Comment by sbrannen** [comment]
FYI: I simplified the provided test case to demonstrate that the bug is neither related to Maven nor specifically to `@EnableRuleMigrationSupport`.

```java
@ExtendWith(ExpectedExceptionSupport.class)
// @RunWith(JUnitPlatform.class) public
class ExpectedExceptionSupportTestCase {

	@Test
	void shouldFail() {
		fail("must fail");
	}

}
```

**Comment by sbrannen** [comment]
Update: changed title accordingly

**Comment by sbrannen** [comment]
Updated _Deliverables_ accordingly.

**Comment by sbrannen** [comment]
_in progress_

**Comment by sbrannen** [comment]
**Analysis**

(_sharing the results of my brainstorming session here_)

The cause of the bug is that `AbstractTestRuleSupport.handleTestExecutionException(ExtensionContext, Throwable)` effectively swallows the exception if there are no registered instances of the `ExpectedException` rule.

My first attempt to fix the bug was to count the number of _members_ that were discovered by having `invokeAppropriateMethodOnRuleAnnotatedMembers()` return the number of members. I then rethrew the exception if the number of members was zero. However, this did not fix the bug since `ExpectedExceptionSupport.handleTestExecutionException(ExtensionContext, Throwable)` invokes `AbstractTestRuleSupport.handleTestExecutionException()` twice: once for fields and once for methods. Since the "member count" does not aggregate across fields + methods, the counting logic is fundamentally broken.

But... to _fix_ the bug, we will in fact have to count the number of _members_ registered; otherwise, we don't know if the adapter is supposed to rethrow the exception.

**Comment by sbrannen** [comment]
FYI: I have pushed my _work in progress_ to the following branch in case anyone wants to pick this up.

https://github.com/junit-team/junit5/tree/issues/1069-expected-exception-support-bug

I'm unassigning myself for the time being.

**Comment by sbrannen** [comment]
The good news is that this is not a show stopper 🐛.

There are two simple workarounds! 😉 

1. If you annotate your test class with `@EnableRuleMigrationSupport`, you then must declare an `ExpectedException` rule even if you don't _use_ it per se -- for example, `@Rule public ExpectedException thrown = ExpectedException.none();`.
2. If you don't want to declare an `ExpectedException` rule that you don't use, then you cannot `@EnableRuleMigrationSupport`. Instead, you can register only the migration support extensions that you actually need -- for example, `@ExtendWith({ExternalResourceSupport.class, VerifierSupport.class})`.

**Comment by sbrannen** [comment]
@algielen, please let us know if the above workarounds work for you.

**Comment by algielen** [comment]
@sbrannen I only need ExternalResourceSupport.class for now. I tested the workaround 2 and it works indeed :)

**Comment by sbrannen** [comment]
Cool. Thanks for the feedback!

**Comment by marcphilipp** [comment]
_in progress_

**Comment by marcphilipp** [comment]
Added another commit to the branch in in a570df8.

@sbrannen Please take another look!

**Comment by sbrannen** [comment]
> Added another commit to the branch in in a570df8.
> 
> @sbrannen Please take another look!

Thanks, @marcphilipp! 

I took a quick glance at the changes in the browser, and I think that's actually the _correct_ way to fix the issue (i.e., making one single pass over all fields and methods that provide rules). So, the changes to the other adapters is likely also for the better (even if they didn't suffer from a bug like this one for expected exceptions).

I'll do my best to look at the code in my IDE tomorrow and then likely merge my test and your fix into `master`.

Ciao 4 now

**Comment by sbrannen** [comment]
Ahhh... I see you left my `numRuleAnnotatedMembers` code in place. So I guess I _partially_ fixed it as well.

Team Work, FTW! 👏

**Comment by sbrannen** [comment]
The _fix_ has been merged into `master` in commit 9de01bafe6439260c5d38a049c253c031cd6f951.

Thanks, @marcphilipp, for the _true_ `@JUnitTeam-Work`. 👍
